patches/server/0860-Damage-with-any-cause.patch

```diff
@@ -0,0 +1,456 @@
++            case FLY_INTO_WALL -> { return DamageSource.FLY_INTO_WALL; }
++            case FREEZE -> { return DamageSource.FREEZE; }
++            case HOT_FLOOR -> { return DamageSource.HOT_FLOOR; }
++            case LAVA -> { return DamageSource.LAVA; }
++            case LIGHTNING -> { return DamageSource.LIGHTNING_BOLT; }
++            case MAGIC -> { return DamageSource.MAGIC; }
++            case MELTING -> { return CraftEventFactory.MELTING; }
++            case POISON -> { return CraftEventFactory.POISON; }
++            case PROJECTILE -> {
++                if (attacker instanceof Projectile projectile) {
++                    EntityType<?> entityType = projectile.getType();
++                    if (entityType == EntityType.TRIDENT) {
++                        return DamageSource.trident(attacker, projectile);
++                    } else if (entityType == EntityType.ARROW || entityType == EntityType.SPECTRAL_ARROW) {
++                        return DamageSource.arrow((Arrow) projectile, projectile.getOwner());
++                    } else if (entityType == EntityType.SNOWBALL || entityType == EntityType.EGG || entityType == EntityType.ENDER_PEARL || entityType == EntityType.POTION) {
++                        return DamageSource.thrown(projectile, projectile.getOwner());
++                    } else if (entityType == EntityType.FIREWORK_ROCKET) {
++                        return DamageSource.fireworks((FireworkRocketEntity) attacker, projectile);
++                    } else if (entityType == EntityType.WITHER_SKULL) {
++                        return DamageSource.witherSkull((WitherSkull) attacker, projectile);
++                    }
++                    return DamageSource.thrown(projectile, projectile.getOwner());
++                }
++                return DamageSource.GENERIC;
++            }
++            case STARVATION -> { return DamageSource.STARVE; }
++            case SUFFOCATION -> { return DamageSource.IN_WALL; }
++            case THORNS -> {
++                if (attacker == null) return DamageSource.GENERIC;
++                return DamageSource.thorns(attacker);
++            }
++            case VOID -> { return DamageSource.OUT_OF_WORLD; }
++            case WITHER -> { return DamageSource.WITHER; }
++            default -> { return DamageSource.GENERIC; }
++        }
++        return DamageSource.GENERIC;
++    }
++}
+diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
+index 4584dcec835f97e94dcf12788b66703075dafe45..5df294a300baa552266adb53da0f6e83088eab31 100644
+--- a/src/main/java/net/minecraft/world/entity/Entity.java
++++ b/src/main/java/net/minecraft/world/entity/Entity.java
+@@ -1874,6 +1874,12 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, i
+     }
+ 
+     public boolean hurt(DamageSource source, float amount) {
++    // Paper start
```

Comment:
Why does this change exist? You never seem to add a call to your new three argument version of `hurt`.

Reply:
It is used at [`CraftEnderDragonPart`](https://github.com/PaperMC/Paper/pull/7423/files#diff-e407df4e89be01749a6a72eed330111c8649ad20abc09e49a8fea72f7f80023fR213) which implements `EnderDragonPart` which extends `Damagable` which extends `Entity`.
```java
// Paper start
@Override
public void damage(double amount, DamageCause cause, Entity source) {
    getParent().damage(amount, cause, source);
}
// Paper end
```
It is used to make EnderDragonPart to be damaged with specific source(entity).

That method is also used at `CraftLivingEntity` too.

I hope I understood the concept of the original PR(#5191)

Reply:
That's `CraftEntity.damage()`, not `Entity.hurt()`. The former calls the two argument version of the latter. I think either the three argument version isn't supposed to exist or that was where it was supposed to get called although the third argument doesn't appear to _do_ anything so I'm leaning toward shouldn't exist.

Reply:
If I understood the definition of the `abstract class`, it is somewhat hybrid of `interface` and actual `class`, thus making actual method (not structure only) can be inherited via implementation.  
`CraftEntity` doesn't override `damage()` from `Entity`, and since `Entity` is abstract class, full `Entity#damage()` will be inherited to `CraftEntity`.

Please correct me if I am wrong.

Reply:
The hurt method hurt(DamageSource, float) was referenced from some non spigot code, which i simply did not want to refactor in my original pr - it was not even needed. 

Therefore i only overloaded the method and made the old method call new overloaded method with null argument, that way the pr do not have to do that many changes all over the place

Reply:
> pr do not have to do that many changes all over the place

Well, that makes me kind of confused. When I first look at the original PR, I thought the codes were nice and clean, no clutterness present, no need to refactor the whole logic of the PR... In this case, I don't have any idea of 'how to make these codes cleaner'.

Any advise is welcomed. I must be misunderstood something wrong.